components/brave_wallet/browser/eth_allowance_manager.cc

```diff
@@ -0,0 +1,409 @@
+    const uint256_t& max_block_number,
+    std::vector<mojom::AllowanceInfoPtr> alwns) {
+  allowances.insert_or_assign(
+      chain_id, std::make_tuple(max_block_number, std::move(alwns)));
+  MarkComplete();
+}
+void EthAllowanceManager::EthAllowanceTask::MarkComplete() {
+  is_completed = true;
+}
+
+EthAllowanceManager::EthAllowanceManager(BraveWalletService* wallet_service,
+                                         JsonRpcService* json_rpc_service,
+                                         KeyringService* keyring_service,
+                                         PrefService* prefs)
+    : wallet_service_(wallet_service),
+      json_rpc_service_(json_rpc_service),
+      keyring_service_(keyring_service),
+      prefs_(prefs) {}
+
+EthAllowanceManager::~EthAllowanceManager() = default;
+
+/**
+ * @brief Represents start of the allowance discovering operation
+ *
+ * @param callback callback to front end, to pass allowances list, error code
+ * and description
+ */
+void EthAllowanceManager::DiscoverEthAllowancesOnAllSupportedChains(
+    ResultCallback callback) {
+  if (!discover_eth_allowance_callback_.empty()) {
+    discover_eth_allowance_callback_.push_back(std::move(callback));
+    return;
+  }
+  discover_eth_allowance_callback_.push_back(std::move(callback));
+
+  const auto keyring_info =
+      keyring_service_->GetKeyringInfoSync(mojom::kDefaultKeyringId);
+  std::vector<std::string> account_addresses;
+  for (const auto& account_info : keyring_info->account_infos) {
+    account_addresses.push_back(account_info->address);
+  }
+
+  if (account_addresses.empty()) {
+    OnDiscoverEthAllowancesCompleted({});
+    return;
+  }
+
+  const auto token_list_map =
+      BlockchainRegistry::GetInstance()->GetEthTokenListMap(
+          GetChainIdsForAlowanceDiscovering());
+
+  base::flat_map<std::string, base::Value::List> chain_id_to_contract_addresses;
+
+  for (const auto& [chain_id, token_list] : token_list_map) {
+    for (const auto& token : token_list) {
+      chain_id_to_contract_addresses[chain_id].Append(token->contract_address);
+    }
+  }
+
+  if (chain_id_to_contract_addresses.empty()) {
+    OnDiscoverEthAllowancesCompleted({});
+    return;
+  }
+
+  const auto& allowance_cache_dict =
+      prefs_->GetDict(kBraveWalletEthAllowancesCache);
+
+  if (!tasks_.empty()) {
+    tasks_.clear();
+  }
+
+  int task_id(0);
+  const auto approval_topic_hash = KeccakHash(kApprovalTopicHash);
+  for (const auto& account_address : account_addresses) {
+    for (const auto& [chain_id, contract_addresses] :
+         chain_id_to_contract_addresses) {
+      const auto* chain_cached_data = allowance_cache_dict.FindDict(chain_id);
+
+      const auto* last_block_number_ptr =
+          chain_cached_data ? chain_cached_data->FindString(kLastBlockNumber)
+                            : nullptr;
+      const std::string last_block_number =
+          last_block_number_ptr != nullptr ? *last_block_number_ptr : "";
```

Comment:
Please move this to anonymous function like GetBlocknumberFilterFromCache, which returns filter string, so code below becomes 
```
if (auto block_number_filter = GetBlocknumberFilterFromCache(...)){
 filter_options.Set(kFromBlock, *block_number_filter);
} else {
 filter_options.Set(kFromBlock, kEarliestBlock);
}
```